omniledger/service/service.go

```diff
@@ -80,163 +85,181 @@ func (s *Service) CreateSkipchain(req *CreateSkipchain) (
+			s.queuesMutex.Unlock()
+			sb, err := s.db().GetLatest(s.db().GetByID(req.SkipchainID))
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			log.Lvlf2("Creating block with transactions %+v", q.transactions)
+			_, err = s.createNewBlock(req.SkipchainID, sb.Roster, q.transactions)
+			if err != nil {
+				log.Error("couldn't create new block: " + err.Error())
+				return
+			}
+		}()
+	} else {
+		s.queues[ssid].transactions = append(s.queues[ssid].transactions, req.Transaction)
+	}
+	return &SetKeyValueResponse{
+		Version:     CurrentVersion,
+		QueueLength: len(s.queues[ssid].transactions),
+	}, nil
+}
+
+// GetProof searches for a key and returns a proof of the
+// presence or the absence of this key.
+func (s *Service) GetProof(req *GetProof) (resp *GetProofResponse, err error) {
+	if req.Version != CurrentVersion {
+		return nil, errors.New("version mismatch")
+	}
+	log.Lvlf2("Getting proof for key %x on sc %x", req.Key, req.ID)
+	latest, err := s.db().GetLatest(s.db().GetByID(req.ID))
 	if err != nil {
-		return nil, errors.New(
-			"Could not get latest block from the skipchain: " + err.Error())
-	}
-	priv := s.storage.Private[gid]
-	if priv == nil {
-		return nil, errors.New("don't have this identity stored")
-	}
-
-	// Verify darc
-	// Note: The verify function needs the collection to be up to date.
-	// TODO: Make sure that is the case.
-	/*
-			log.Lvl1("Verifying signature")
-		    err := s.getCollection(req.SkipchainID).verify(&req.Transaction)
-		    if err != nil {
-				log.Lvl1("signature verification failed")
-		        return nil, err
-		    }
-			log.Lvl1("signature verification succeeded")
-	*/
-
-	coll := s.getCollection(req.SkipchainID)
-	if _, _, err := coll.GetValue(req.Transaction.Key); err == nil {
-		return nil, errors.New("cannot overwrite existing value")
-	}
-	sigBuf, err := network.Marshal(&req.Transaction.Signature)
+		return
+	}
+	proof, err := NewProof(s.getCollection(req.ID), s.db(), latest.Hash, req.Key)
 	if err != nil {
-		return nil, errors.New("Couldn't marshal Signature: " + err.Error())
+		return
+	}
+	resp = &GetProofResponse{
+		Version: CurrentVersion,
+		Proof:   *proof,
+	}
+	return
+}
+
+// createNewBlock creates a new block and proposes it to the
+// skipchain-service. Once the block has been created, we
+// inform all nodes to update their internal collections
+// to include the new transactions.
+func (s *Service) createNewBlock(scID skipchain.SkipBlockID, r *onet.Roster, ts []Transaction) (*skipchain.SkipBlock, error) {
+	var sb *skipchain.SkipBlock
+	var c collection.Collection
+
+	if scID.IsNull() {
+		// For a genesis block, we create a throwaway collection.
+		c = collection.New(&collection.Data{}, &collection.Data{})
+
+		sb = skipchain.NewSkipBlock()
+		sb.Roster = r
+		sb.MaximumHeight = 10
+		sb.BaseHeight = 10
+	} else {
+		// For further blocks, we create a clone of the collection - this is
+		// TODO: not very memory-friendly - we need to use some kind of transactions.
```

Comment:
+1 I prefer transactions over the try-to-write-then-delete method, but we need to implement something new in that case, perhaps [WAL](https://en.wikipedia.org/wiki/Write-ahead_logging)

Reply:
@pablo-lo created a new method for it, but it's also write-then-delete. I'll ask Raphael if there is something better.